File: examples/oracle.ml
※ The Oracle — a concurrent MapReduce analytics engine you can question.
※
※ Phase 1, ingest: strand 0 reads document lines until «.», dealing them
※ onto channel δ, then deals one ∅ per mapper. Strands 1–3 are mappers:
※ three receivers on one channel form a work-stealing pool — each steals
※ a line, lowercases and splits it, and streams the words to the counter
※ channel ω, signing off with its own ∅ when the deck runs dry.
※
※ Phase 2, serve: strand 4 folds ω into an association list until all
※ three mappers have signed off, then becomes the state-owning actor:
※ requests arrive on κ, answers leave on β. Strand 5 opens the floor
※ (after the go-signal γ — it must not race strand 0 for stdin), reads
※ one query per line, and parses each on a freshly spawned strand inside
※ ⍥: a malformed command reports «✗ …» and dies alone, the calc.ml
※ let-it-crash pattern. Strand 6 prints every answer, in command order
※ because strand 5 joins each parser before reading on.
※
※   count WORD · top K · stats · q     (document first, closed by «.»)
※   printf 'There is no spoon⏎.⏎count spoon⏎top 3⏎stats⏎q⏎' \
※     | mlang run examples/oracle.ml
※
※ B: counts word → counts′ — bump one word's tally.
[⇒z ∂[⊃z=]⌿⇒m [⊃z≠]⌿ ⟨⟨z m⟨⟩=[1][m⊃1@1+]?⟩⟩⧺]≔B
※ G: counts word → n — look a tally up, 0 when the Oracle never saw it.
[⇒z [⊃z=]⌿⇒m m⟨⟩=[0][m⊃1@]?]≔G
※ T: counts k → report — top k by count, ties alphabetical: each pair
※ becomes the sortable key «(11000−n) word» (counts stay below 1000),
※ so one string sort ⍋ orders by count descending, then word ascending.
[⇒k [∂1@11000⇅-⍕« »⧺⇅⊃⧺]∵⍋ k⊤ [⍭∂1@⇅0@⍎11000⇅-⍕« × »⇅⧺⧺]∵«⏎»⊇]≔T
⇊
0⇒n [⌨∂∅≠⊚«.»≠∧][↥δ n1+⇒n]⟳⌫ n≔Y 3[∅↥δ]⍣
[↧δ∂∅≠][⇩⍭[↥ω]∀]⟳⌫ ∅↥ω
[↧δ∂∅≠][⇩⍭[↥ω]∀]⟳⌫ ∅↥ω
[↧δ∂∅≠][⇩⍭[↥ω]∀]⟳⌫ ∅↥ω
0⇒d ⟨⟩ [d3<][↧ω∂∅=[⌫d1+⇒d][B]?]⟳ 1↥γ 1⇒g
⋮ [g][↧κ⇒q q⊃«count»=[∂q1@G q1@«: »⧺⇅⍕⧺↥β][q⊃«top»=[∂q1@T↥β][q⊃«stats»=[∂[1@]∵∑⇒w «lines »Y⍕⧺« · words »⧺w⍕⧺« · distinct »⧺⊚#⍕⧺↥β][0⇒g «goodbye, operator»↥α ∅↥α]?]?]?]⟳⌫
↧γ⌫ 1⇒r [r][⌨⇒v v∅=v«q»=∨[0⇒r ⟨«halt»⟩↥κ][[[v⍭⇒t t⊃«count»=[⟨«count»t1@⟩][t⊃«top»=[⟨«top»t1@⍎⟩][t⊃«stats»=[⟨«stats»⟩][«unknown command»↯]?]?]?↥κ ↧β↥α][⌫«✗ »v⧺↥α]⍥]⚡⋈]?]⟳
[↧α∂∅≠][⍞]⟳⌫

File: bench/python_ports/oracle.py
import queue
import sys
import threading

MAPPERS = 3
lines_q = queue.Queue()
words_q = queue.Queue()
requests = queue.Queue()
replies = queue.Queue()
answers = queue.Queue()

doc_lines = 0


def mapper():
    while True:
        line = lines_q.get()
        if line is None:
            words_q.put(None)
            return
        for word in line.lower().split():
            words_q.put(word)


def counter():
    counts = {}
    done = 0
    while done < MAPPERS:
        w = words_q.get()
        if w is None:
            done += 1
        else:
            counts[w] = counts.get(w, 0) + 1
    while True:
        req = requests.get()
        if req[0] == "count":
            word = req[1]
            replies.put(f"{word}: {counts.get(word, 0)}")
        elif req[0] == "top":
            ranked = sorted(counts.items(), key=lambda kv: (-kv[1], kv[0]))
            replies.put("\n".join(f"{w} × {n}" for w, n in ranked[:req[1]]))
        elif req[0] == "stats":
            total = sum(counts.values())
            replies.put(f"lines {doc_lines} · words {total} "
                        f"· distinct {len(counts)}")
        else:
            answers.put("goodbye, operator")
            answers.put(None)
            return


def printer():
    while True:
        a = answers.get()
        if a is None:
            return
        print(a)


def handle(line):
    try:
        parts = line.split()
        cmd = parts[0]
        if cmd == "count":
            requests.put(("count", parts[1]))
        elif cmd == "top":
            requests.put(("top", int(parts[1])))
        elif cmd == "stats":
            requests.put(("stats",))
        else:
            raise ValueError("unknown command")
        answers.put(replies.get())
    except Exception:
        answers.put("✗ " + line)


threads = [threading.Thread(target=mapper) for _ in range(MAPPERS)]
threads += [threading.Thread(target=counter), threading.Thread(target=printer)]
for t in threads:
    t.start()

stdin = (line.rstrip("\n") for line in sys.stdin)
for line in stdin:
    if line == ".":
        break
    lines_q.put(line)
    doc_lines += 1
for _ in range(MAPPERS):
    lines_q.put(None)

for line in stdin:
    if line == "q":
        break
    worker = threading.Thread(target=handle, args=(line,))
    worker.start()
    worker.join()
requests.put(("halt",))
for t in threads:
    t.join()
